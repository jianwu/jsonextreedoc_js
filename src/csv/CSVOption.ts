export class CSVOption {
  public includeHeader = true;
  public fieldSep = ',';
  public recordSep = '\n';
  public quoteChar = '"';

  public _fieldAndRecord = this.fieldSep + this.recordSep;

  public setFieldSep(fieldSep: string) { this.fieldSep = fieldSep; return this.buildTerm(); }
  public setRecordSep(recordSep: string) { this.recordSep = recordSep; return this.buildTerm(); }
  public setQuoteChar(quoteChar: string) { this.quoteChar = quoteChar; return this; }
  public setIncludeHeader(includeHeader: boolean) { this.includeHeader = includeHeader; return this; }

  public buildTerm() {
    this._fieldAndRecord = this.fieldSep + this.recordSep;
    return this;
  }
}
