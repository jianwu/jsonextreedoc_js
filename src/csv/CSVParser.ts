import TDNode, { TDNodeType } from '../TDNode';
import CSVOption from './CSVOption';
import CharSource from '../core/CharSource';
import StringBuilder from '../core/StringBuilder';
import TreeDoc from '../TreeDoc';
import { StringCharSource } from '..';

const SPACE_CHARS = " \r";

export default class CSVParser {

  public static readonly instance = new CSVParser();
  public static get() { return CSVParser.instance; }

  public static parse(src: CharSource | string, opt: CSVOption): TDNode {
    return CSVParser.get().parse(src, opt);
  }

  public parse(src: CharSource | string, opt: CSVOption = new CSVOption(), root = new TreeDoc('root').root): TDNode {
    if (typeof src === 'string')
      src = new StringCharSource(src);

    root.setType(TDNodeType.ARRAY);
    while (!src.isEof()) {
      if (!src.skipChars(SPACE_CHARS))
        break;
      this.readRecord(src, opt, root);
    }
    return root;
  }

  readRecord(src: CharSource, opt: CSVOption, root: TDNode) {
    const row = new TDNode(root.doc);
    row.setStart(src.getBookmark());
    while(!src.isEof() && src.peek() !== opt.recordSep) {
      if (!src.skipChars(SPACE_CHARS))
        break;
      const start = src.getBookmark();
      const fieldNode = row.createChild().setType(TDNodeType.ARRAY).setValue(this.readField(src, opt));
      fieldNode.setStart(start).setEnd(src.getBookmark());
    }
    row.setEnd(src.getBookmark());
    if (row.getChildrenSize() > 0)
      root.addChild(row);
    if (!src.isEof())
      src.read();  // Skip the recordSep
  }

  readField(source: CharSource, opt: CSVOption): string {
    const sb = new StringBuilder();
    let previousQuoted = false;
    while(!source.isEof() && source.peek() !== opt.fieldSep && source.peek() !== opt.recordSep) {
      if (source.peek() === opt.quoteChar) {
        if (previousQuoted)
          sb.append(opt.quoteChar);
        source.skip();  // for "", we will keep one quote
        source.readUntilTerminatorToString(opt.quoteChar, sb);
        if (source.peek() === opt.quoteChar)
          source.skip();
        previousQuoted = true;
      } else {
        sb.append(source.readUntilTerminator(opt.fieldSep + opt.recordSep).trim());
        previousQuoted = false;
      }
    }
    if (!source.isEof() && source.peek() === opt.fieldSep)
      source.skip();  // Skip fieldSep

    return sb.toString();
  }
}
