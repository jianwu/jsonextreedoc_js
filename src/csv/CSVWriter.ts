import CSVOption from './CSVOption';
import TDNode, { TDNodeType } from '../TDNode';
import StringBuilder from '../core/StringBuilder';
import Appendable from '../core/Appendable';
import StringUtil from '../core/StringUtil';
import { TreeDoc } from '..';

const { contains } = StringUtil;

export default class CSVNWriter {
  public static readonly instance = new CSVNWriter();
  public static get() {
    return CSVNWriter.instance;
  }

  public writeAsString(node: TDNode, opt = new CSVOption()): string { return this.write(new StringBuilder(), node, opt).toString(); }

  public write(out: Appendable, node: TDNode, opt: CSVOption): typeof out {
    if (node.children) {
      for (const row of node.children) {
        if (row.children) {
          for (const field of row.children) {
            this.writeField(out, field, opt);
            out.append(opt.fieldSep);
          }
          out.append(opt.recordSep);
        }
      }
    }
    return out;
  }

  writeField(out: Appendable, field: TDNode, opt: CSVOption): typeof out {
    const quote = opt.quoteChar;
    let str = "" + field.value;
    if (contains(str, quote) || contains(str, opt.fieldSep) || contains(str, opt.recordSep)) {
      if (contains(str, quote))
        str = str.replace(new RegExp(quote, 'g'), quote + quote);
      return out.append(quote).append(str).append(quote);
    }
    return out.append(str);
  }
}
