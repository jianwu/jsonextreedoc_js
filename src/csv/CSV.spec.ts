import { CSVParser } from './CSVParser';
import { CSVOption } from './CSVOption';
import { CSVWriter } from './CSVWriter';
import { StringCharSource, TDJSONParser } from '..';

const testCsv = `
field1,field2,field3,field4
v11,v12,"",1
v21, "v2l1,
V2l2" ,v23,true
"v31""v31","v32""""v32",v33,"3"
`;

const testObj = `
"@key","field1","field2"
k1,v11,v12
k2,v21,v22
`

const csvWithJson = `
val,count,percent
"[{""kind"":""exact"",""field_path"":""k8s_environment"",""value"":""production""}]",73,0.24333333333333335
"[{""kind"":""exact"",""field_path"":""k8s_environment"",""value"":""production""},{""kind"":""regex"",""field_path"":""k8s_namespace""}]",61,0.20333333333333334
`

describe('CSVParser and CSVWriter', () => {
  function testParseAndWrite(opt: CSVOption, csv: string) {
    const node = CSVParser.get().parse(csv, opt);
    expect(node.toString()).toMatchSnapshot();
    const str = CSVWriter.get().writeAsString(node, opt.setFieldSep('|'));
    expect(str).toMatchSnapshot();
    const node1 = CSVParser.get().parse(str, opt);
    expect(node1.toString()).toBe(node.toString());
  }

  test('ParseAndWriteWithoutHeader', () => testParseAndWrite(new CSVOption().setIncludeHeader(false), testCsv));
  test('ParseAndWriteWithHeader', () => testParseAndWrite(new CSVOption(), testCsv));
  test('ParseAndWriteObj', () => testParseAndWrite(new CSVOption(), testObj));
  
  test('testParseAndWriteJson', () => testParseAndWrite(new CSVOption(), csvWithJson));

  test('JSONValue', () => {
    const json = "[{f1: v1, f2: {a: 1}}, {f1: 0, f3:''}]";
    expect(CSVWriter.get().writeAsString(TDJSONParser.get().parse(json))).toMatchSnapshot();
  });

  test('testReadField', () => {
    expect(CSVParser.get().readField(new StringCharSource("'ab''cd'"), new CSVOption().setQuoteChar("'"))).toBe(
      "ab'cd",
    );
  });

  test('testReadFieldMissingQuote', () => {
    let error = '';
    try {
      CSVParser.get().readField(new StringCharSource("'ab''cd"), new CSVOption().setQuoteChar("'"));
    } catch (e: any) {
      error = e.message;
    }
    expect(error).toMatchInlineSnapshot(
      `"Can't find matching quote at position:5;line:0;col:5, Bookmark(line=0, col=7, pos=7), digest:"`,
    );
  });
});
