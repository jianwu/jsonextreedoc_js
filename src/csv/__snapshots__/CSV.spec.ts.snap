// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CSVParser and CSVWriter testParseAndWriter 1`] = `"[['field1', 'field2', 'field3', 'field4'], ['v11', 'v12', 'v13', 1], ['v21', 'v2l1\\\\nV2l2', 'v23', true], ['v31\\"v31', 'v32\\"\\"v32', 'v33', '3']]"`;

exports[`CSVParser and CSVWriter testParseAndWriter 2`] = `
"field1|field2|field3|field4|
v11|v12|v13|1|
v21|\\"v2l1
V2l2\\"|v23|true|
\\"v31\\"\\"v31\\"|\\"v32\\"\\"\\"\\"v32\\"|v33|\\"3\\"|
"
`;

exports[`CSVParser and CSVWriter testParseAndWriter 3`] = `"[['field1', 'field2', 'field3', 'field4'], ['v11', 'v12', 'v13', 1], ['v21', 'v2l1\\\\nV2l2', 'v23', true], ['v31\\"v31', 'v32\\"\\"v32', 'v33', '3']]"`;
